x-pack/filebeat/input/azureeventhub/v2_input.go

```diff
@@ -0,0 +1,532 @@
+			ClientOptions: azcore.ClientOptions{
+				Cloud: cloud.AzurePublic,
+			},
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create blob container pipelineClient: %w", err)
+	}
+
+	// The modern event hub SDK does not create the container
+	// automatically like the old SDK.
+	//
+	// The new `BlobStore` explicitly says:
+	//   "the container must exist before the checkpoint store can be used."
+	//
+	// We need to ensure it exists before we can use it.
+	err = in.ensureContainerExists(ctx, containerClient)
+	if err != nil {
+		return fmt.Errorf("failed to ensure blob container exists: %w", err)
+	}
+
+	// The checkpoint store is used to store the checkpoint information
+	// in the blob container.
+	checkpointStore, err := checkpoints.NewBlobStore(containerClient, nil)
+	if err != nil {
+		return fmt.Errorf("failed to create checkpoint store: %w", err)
+	}
+	in.checkpointStore = checkpointStore
+
+	// Create the event hub consumerClient to receive events.
+	consumerClient, err := azeventhubs.NewConsumerClientFromConnectionString(
+		in.config.ConnectionString,
+		in.config.EventHubName,
+		in.config.ConsumerGroup,
+		nil,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create consumer pipelineClient: %w", err)
+	}
+	in.consumerClient = consumerClient
+
+	// Manage the migration of the checkpoint information
+	// from the old Event Hub SDK to the new Event Hub SDK.
+	in.migrationAssistant = newMigrationAssistant(
+		in.log,
+		consumerClient,
+		containerClient,
+		checkpointStore,
+	)
+
+	return nil
+}
+
+// run starts the main loop for processing events.
+func (in *eventHubInputV2) run(ctx context.Context) {
+	if in.config.MigrateCheckpoint {
+		in.log.Infow("checkpoint migration is enabled")
+		// Check if we need to migrate the checkpoint store.
+		err := in.migrationAssistant.checkAndMigrate(
+			ctx,
+			in.config.ConnectionString,
+			in.config.EventHubName,
+			in.config.ConsumerGroup,
+		)
+		if err != nil {
+			in.log.Errorw("error migrating checkpoint store", "error", err)
+			// FIXME: should we return here?
+		}
+	}
+
+	// Handle the case when the processor stops due to
+	// transient errors (network failures) and we need to
+	// restart it.
+	processorRunBackoff := backoff.NewEqualJitterBackoff(
+		ctx.Done(),
+		processorRestartBackoff,    // initial backoff
+		processorRestartMaxBackoff, // max backoff
+	)
+
+	// Create the processor options using the input
+	// configuration.
+	processorOptions := createProcessorOptions(in.config)
+
+	for ctx.Err() == nil {
+		// Create a new processor for each run.
+		//
+		// The docs explicitly say that the processor
+		// is not reusable.
+		processor, err := azeventhubs.NewProcessor(
+			in.consumerClient,
+			in.checkpointStore,
+			processorOptions,
+		)
+		if err != nil {
+			in.log.Errorw("error creating processor", "error", err)
+			return
+		}
+
+		// Launch one goroutines for each partition
+		// to process events.
+		go in.workersLoop(ctx, processor)
```

Comment:
The goroutine started for each partition doesn't have a way to signal when it's done. This could potentially lead to resource leaks if the main loop exits before all goroutines have finished.

Use waitgroups or [errgroup](golang.org/x/sync/errgroup)? And before logging "run completed"; wait for all spawned goroutines to return?

Also, do we this spawn a goroutine for in.workersLoop; inside workersLoop, anyway we are spawning goroutines?

Reply:
Example: https://sourcegraph.com/github.com/tsaikd/gogstash/-/blob/input/azureeventhub/inputazureeventhub.go?L184

Reply:
Got it.